### DISCUSSION TITLE
`destruct` expression

### DISCUSSION THREAD

**Discussion by mlutze** [flix/flix#7669] [Ideas]
Very often we write a function that looks like this:

```scala
def visit(e: Expr): Bool = match e {
    case Add(e1, e2) => foo(e1); foo(e2); add(visit(e), visit(e2))       // oops
    case Sub(e1, e2) => ...
}
```

A common bug for these cases is that we accidentally use `e` where we should have used `e1` or `e2`. I want to prevent this kind of bug by removing `e` from the environment when I match on it. I propose a new expression `destruct`, which operates only on variables, and performs a match, but removes the scrutinee from the environment.

```scala
def visit(e: Expr): Bool = match e {
    case Add(e1, e2) => foo(e1); foo(e2); add(visit(e), visit(e2))       // compile error: can't find `e`. logic error caught!
    case Sub(e1, e2) => ...
}
```

**Comment by magnus-madsen** [comment]
Seems reasonable, but I am not 100% sure. We can look into it once the new parser is merged and stable.

**Comment by JonathanStarup** [comment]
One important example with this system is:
```
def visit(e: Expr): Bool = destruct e {
    case Add(e1, e2) => let e = e1; foo(e); foo(e2); add(visit(e), visit(e2))
    case Sub(e1, e2) => ...
}
```
I think that `destruct` should not remove `e` per se, but just mark it illegal to use. this makes shadowing like the above an error

A related idea to this is explicit lambda capture, so that you can avoid using outer variables in inner functions
